x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/bind/BindRequestBuilder.java

```diff
@@ -0,0 +1,154 @@
+import static org.elasticsearch.xpack.core.security.authc.ldap.PoolingSessionFactorySettings.BIND_MODE;
+import static org.elasticsearch.xpack.core.security.authc.ldap.PoolingSessionFactorySettings.LEGACY_BIND_PASSWORD;
+import static org.elasticsearch.xpack.core.security.authc.ldap.PoolingSessionFactorySettings.SASL_GSSAPI_DEBUG;
+import static org.elasticsearch.xpack.core.security.authc.ldap.PoolingSessionFactorySettings.SASL_GSSAPI_KEYTAB_PATH;
+import static org.elasticsearch.xpack.core.security.authc.ldap.PoolingSessionFactorySettings.SASL_GSSAPI_PRINCIPAL;
+import static org.elasticsearch.xpack.core.security.authc.ldap.PoolingSessionFactorySettings.SECURE_BIND_PASSWORD;
+
+/**
+ * Supports building bind request from given configuration either a {@link SimpleBindRequest} or {@link SASLBindRequest}
+ */
+public final class BindRequestBuilder {
+    private final RealmConfig realmConfig;
+    private final Function<RealmConfig, String> extractBindDn;
+
+    public BindRequestBuilder(final RealmConfig realmConfig, final Function<RealmConfig, String> extractBindDn) {
+        this.realmConfig = realmConfig;
+        this.extractBindDn = extractBindDn;
+    }
+
+    public BindRequest build() throws LDAPException {
+        final BindRequest bindRequest;
+        final String mode = realmConfig.getSetting(BIND_MODE);
+        final byte[] bindPassword = bindPassword(realmConfig);
+        switch (mode) {
+        case "simple":
+            bindRequest = buildSimpleBindRequest(extractBindDn.apply(realmConfig), bindPassword);
+            break;
+        case "sasl_gssapi":
+            bindRequest = buildGSSAPIBindRequest(bindPassword);
+            break;
+        default:
+            throw new IllegalArgumentException("unsupported bind request mode, valid modes ['simple', 'sasl_gssapi']");
+        }
+        return bindRequest;
+    }
+
+    private BindRequest buildGSSAPIBindRequest(final byte[] bindPassword) throws LDAPException {
+        final BindRequest bindRequest;
+        Path keytabPath = validateGSSAPISettings(bindPassword);
+
+        final String principal = realmConfig.getSetting(SASL_GSSAPI_PRINCIPAL);
+        final GSSAPIBindRequestProperties gssapiBindRequestProperties = new GSSAPIBindRequestProperties(principal, bindPassword);
+        if (keytabPath != null) {
+            gssapiBindRequestProperties.setUseKeyTab(true);
+            gssapiBindRequestProperties.setKeyTabPath(keytabPath.toString());
+        }
+        if (realmConfig.hasSetting(SASL_GSSAPI_DEBUG)) {
+            gssapiBindRequestProperties.setEnableGSSAPIDebugging(realmConfig.getSetting(SASL_GSSAPI_DEBUG));
+        }
+        gssapiBindRequestProperties.setRefreshKrb5Config(true);
+        gssapiBindRequestProperties.setUseTicketCache(false);
+        bindRequest = new GSSAPIBindRequest(gssapiBindRequestProperties);
+        return bindRequest;
+    }
+
+    private Path validateGSSAPISettings(final byte[] bindPassword) {
```

Comment:
Call me hypercritical, but this is woefully complex. And it all stems from that conditional setting I'm gripping about...

What I don't lilke:
This class can deduce the SASL principal, _if it needs to_, but it cannot deduce the bind DN. So, in case it needs the bind DN, it will use the provided function for that. But the caller needs to know when the bind DN is required, so that it can pass the function. This requires for the caller to know of the internals of the `BindRequestBuilder` which is a violation of the abstraction.

Reply:
I suggest we create a separate `BindRequestBuilder` for the SASL case only, and have the caller (or another wrapper) handle the decision of which `BindRequest` type to build, given the realm settings.